stan/math/prim/fun/columns_dot_self.hpp

```diff
@@ -9,17 +10,16 @@ namespace math {
 /**
  * Returns the dot product of each column of a matrix with itself.
  *
- * @tparam T type of elements in the matrix
- * @tparam R number of rows, can be Eigen::Dynamic
- * @tparam C number of columns, can be Eigen::Dynamic
+ * @tparam T type of the matrix
  *
  * @param x Matrix.
  * @return Vector containing the dot product of each column of the matrix
  * with itself.
  */
-template <typename T, int R, int C>
-inline Eigen::Matrix<T, 1, C> columns_dot_self(
-    const Eigen::Matrix<T, R, C>& x) {
+template <typename T, require_eigen_t<T>* = nullptr,
+          require_not_eigen_vt<is_var, T>* = nullptr>
+inline Eigen::Matrix<value_type_t<T>, 1, T::ColsAtCompileTime> columns_dot_self(
```

Comment:
One reason I'd like to have auto on these is that we can add .eval() or .matrix() now and once we get these all working with generic eigen types we can just remove the .eval() and let it all rock and roll

Reply:
This is again a case where eval returns a row major matrix.